ipn/ipnauth/access_check.go

```diff
@@ -0,0 +1,346 @@
+//
+// Higher-level access control code should forward the AccessCheckResult
+// from any lower-level access check to the caller as soon as it receives
+// a definitive result as indicated by the HasResult() method returning true.
+// Therefore, if access is denied due to tailscaled config or syspolicy settings,
+// it will be immediately denied, regardless of the caller's identity.
+func DenyAccess(err error) AccessCheckResult {
+	if err == nil {
+		err = ipn.NewInternalServerError("access denied with a nil error")
+	} else {
+		err = &ipn.AccessDeniedError{Err: err}
+	}
+	return AccessCheckResult{err: err, hasResult: true}
+}
+
+// ContinueCheck returns a new AccessCheckResult indicating that
+// the requested access has neither been allowed, nor denied,
+// and any further access checks should be performed to determine the result.
+//
+// An an example, a higher level access control code that denies
+// certain access rights based on syspolicy may return ContinueCheck()
+// to indicate that access is not denied by any applicable policies,
+// and lower-level access checks should be performed.
+//
+// Similarly, if a tailscaled config file is present and restricts certain ipn.Prefs fields
+// from being modified, its access checking mechanism should return ContinueCheck()
+// when a user tries to change only preferences that are not locked down.
+//
+// As a general rule, any higher-level access checking code should
+// continue calling lower-level access checking code, until it either receives
+// and forwards a definitive result from one of the lower-level mechanisms,
+// or until there are no additional checks to be performed.
+// In the latter case, it can also return ContinueCheck(),
+// resulting in the requested access being implicitly denied.
+func ContinueCheck() AccessCheckResult {
+	return AccessCheckResult{}
+}
+
+// HasResult indicates if a definitive access decision (either allowed or denied) has been made.
+func (r AccessCheckResult) HasResult() bool {
+	return r.hasResult
+}
+
+// Allowed reports whether the requested access has been allowed.
+func (r AccessCheckResult) Allowed() bool {
+	return r.hasResult && r.err == nil
+}
+
+// Denied reports whether the requested access should be denied.
+func (r AccessCheckResult) Denied() bool {
+	return !r.hasResult || r.err != nil
+}
+
+// Error returns an ipn.AccessDeniedError detailing why access was denied,
+// or nil if access has been allowed.
+func (r AccessCheckResult) Error() error {
+	if !r.hasResult && r.err == nil {
+		return errNotAllowed
+	}
+	return r.err
+}
+
+// String returns a string representation of r.
+func (r AccessCheckResult) String() string {
+	switch {
+	case !r.hasResult:
+		return "Implicit Deny"
+	case r.err != nil:
+		return "Deny: " + r.err.Error()
+	default:
+		return "Allow"
+	}
+}
+
+// accessChecker is a helper type that allows step-by-step granting or denying of access rights.
+type accessChecker[T ~uint32] struct {
+	requested T
```

Comment:
this name sounds like it's immutable once constructed.

maybe two fields, requested and remain? and doc which is if any is immutable